app-ios/tutanota/Sources/Contacts/ContactsSynchronization.swift

```diff
@@ -22,281 +29,259 @@ private let ALL_SUPPORTED_CONTACT_KEYS: [CNKeyDescriptor] = [
-    nativeContacts.merge(storedNativeContacts) { a, _ in a }
-
-    // missing some?
-    for contact in contactsNotFound {
-      nativeContacts[contact] = NativeMutableContact(newContactWithId: contact, container: self.localContainer)
-    }
-
-    // update contact data
-    for contact in contacts {
-      nativeContacts[contact.id]!.updateContactWithData(contact)
-    }
-
-    // finally save
-    try saveNativeContacts(nativeContacts, forUserId: userId, deleteOldContacts: delete)
   }
-
-  private func getSavedLocalContacts(forUserId userId: String, withStructuredContacts structuredContacts: [StructuredContact], deleteOldContacts: Bool) throws -> TutaToNativeContactIdentifiers {
-    // do nothing if we have no contacts stored
-    var localContacts = getSavedLocalContactsIdentifierMappings(forUserId: userId)
-    if localContacts.isEmpty {
-      return [:]
-    }
-
-    // remove everything (likely user deletion, or account has no contacts yet), so no need to loop
-    if structuredContacts.isEmpty {
-      if deleteOldContacts {
-        try removeNativeContacts(localContacts, forUserId: userId)
-      }
-      return [:]
-    }
-
-    // Load into a hashmap for efficient searching
-    let structuredContactsMap: TutaToStructuredContacts = structuredContacts.reduce(into: [:]) { result, contact in
-      result[contact.id] = contact
+  
+  /// Query the local container, ignoring the user's choices for the default contacts location.
+  /// This prevent other apps, as Gmail or even iCloud, from 'stealing' and moving our contacts to their lists.
+  private lazy var localContainer: String = {
+    let store = CNContactStore()
+    let defaultContainer = store.defaultContainerIdentifier()
+    
+    do {
+      let containers = try store.containers(matching: nil)
+      
+      // Apple allow just ONE local container, so we can query for the first and unique one
+      let localContainer = containers.first(where: {$0.type == CNContainerType.local})
+      
+      return localContainer?.identifier ?? defaultContainer
+    } catch {
+      TUTSLog("Failed to get local container, using default")
+      return defaultContainer
     }
-
-    // find everything we have stored that isn't in the new list
-    var contactsToRemove: TutaToNativeContactIdentifiers = [:]
-    for i in localContacts {
-      if structuredContactsMap[i.key] == nil {
-        contactsToRemove[i.key] = i.value
-      }
+  }()
+  
+  private func insert(contacts: [StructuredContact], forUser user: inout UserContactMapping) throws {
+    let store = CNContactStore()
+    let saveRequest = CNSaveRequest()
+    let contactGroup = try getTutaContactGroup(forUser: &user)
+    
+    // We need store mapping from our contact id to native contact id but we get it only after actually saving the contacts,
+    // so until we execute the save request we keep track of the mapping
+    var tutaContactToNativeContact = [String : NativeMutableContact]()
```

Comment:
You should use the TutaToNativeContacts typealias here. It's defined at the top of the file.

```swift
private typealias TutaToNativeContacts = [String: NativeMutableContact]
```

You should be able to initialize it like this:

```swift
var tutaContactToNativeContact = TutaToNativeContacts()
```